### ISSUE TITLE
Send seems to close the websocket ?

### ISSUE THREAD

**Issue by surt666** [hamza1311/reqwasm#13]
Got the send functionality working per the example in the futures::websocket source, but it seems to close the websocket after send, and it seems wrong to do an open each time you send. Shouldn't the websocket be longlived, and is there a way to make it reconnect, if state is Closed or Closing ?. I might just not understand how this crate is supposed to be used.

**Comment by surt666** [comment]
I can try. Background is I'm using Sycamore, so have components where I want to send messages, but create the websocket in the main app component, and send it as a property to underlying components. So have a definition of open and send as this 

`pub fn get_websocket() -> WebSocket {
    WebSocket::open(WS_URL).unwrap()
}

pub fn send_msg(mut ws: WebSocket, msg: String) {
    spawn_local({
        async move {
            ws.send(Message::Text(msg)).await.unwrap();
        }
    })
}`

These are used from other modules. first in the main app as 

`fn main() {
    console_error_panic_hook::set_once();
    console_log::init_with_level(log::Level::Debug).unwrap();
    let ws = get_websocket();
    sycamore::render(|| template! { App(AppProps{ws: ws}) });
}`

Then in the router ws gets send further down with a clone, which might be the issue, not sure.

`AppRoutes::Users => template! {
           Users(pages::users::Props {ws: props.ws.clone()})
},`

And in the User component is used inside a buttons onclick event as

`button(class="col-span-4 bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded justify-self-center", on:click= move |_| {
                    let u = HashMap::from([("user".to_string(), user.get().to_string()), ("name".to_string(),  name.get().to_string())]);
                    let event = Event{service: "command".to_string(), source: "kedl.frontend".to_string(), detail_type: "Create User".to_string(), detail: serde_json::to_string(&u).unwrap()};
                    let json = serde_json::to_string(&event).unwrap();
                    log::info!("U WS STATE {:#?}",props.ws.state());
                    send_msg(props.ws.clone(), json);
                    log::info!("U WS STATE 2 {:#?}",props.ws.state());
            }) { "Create" }`

First time the button is clicked everything works fine and the state is Open, but the second time the state is Closed, in both log statements. 

Don't know if this is possible to make sense off

**Comment by hamza1311** [comment]
`WebSocket` code was re-written in #4 but that isn't released yet (I thought I released it lol). Can you try depending upon reqwasm from `git`? Please let me know how the experience is. I shall release it based on your feedback.

You can look at this very simple test as an example:
https://github.com/hamza1311/reqwasm/blob/fb356775a579cd2cbad1111bcb77fa68c0b83366/src/websocket/futures.rs#L307-L319

**Comment by surt666** [comment]
I was already depending on git, but it still doesn't work for me. If I do 2 sends right after each other (in code, not by clicking a button), the second fails because the socket is closing. I'm not sure how to use the StreamExt::split, it seems to mandate a Context with some Waker stuff, that I have no idea how to do. Still using the  Sink directly, you seem to forego the purpose of having a wrapper around the whole websys::websocket thing, or maybe i'm misunderstanding it totally. What is the correct way to create a socket in one place of your wasm app, and then reuse that in all it's sub components ?. I don't think doing as in the example is viable for real apps with multiple pages and components, but again I might be totally wrong on the use case here.

**Comment by hamza1311** [comment]
I see. This seems to be a bug. The `Clone` and `Drop` implementations should not exist.

> I'm not sure how to use the StreamExt::split, it seems to mandate a Context with some Waker stuff, that I have no idea how to do.

Sorry, I explained it badly. The split method can be used like this:
```rust
use futures::stream::StreamExt;
use wasm_bindgen_futures::spawn_local;

let ws = WebSocket::open(URL)?;

let (mut sender, mut receiver) = ws.split();

spawn_local(async move {
    sender.send(Message::Text(String::from("something"))).await;
});

spawn_local(async move {
    while let Some(msg) = receiver.next() {
        log!(format!("{:?}", msg))
    }
});
```

> Still using the Sink directly, you seem to forego the purpose of having a wrapper around the whole websys::websocket thing, or maybe i'm misunderstanding it totally. What is the correct way to create a socket in one place of your wasm app, and then reuse that in all it's sub components ?. I don't think doing as in the example is viable for real apps with multiple pages and components, but again I might be totally wrong on the use case here.

This problem is exactly why I initially held off the release (then life got in the way and I forgot). This wrapper is based around Rust's Futures. It seems like splitting the future into Sink and Stream and passing those individually seems like the right approach. 

You'd want to have some sort of communication mechanism in your application where a `WebSocket` instance is created at the top of the components tree and the messages are passed around using something like channels (see [`futures::channel::mpsc`](https://docs.rs/futures/0.3.17/futures/channel/mpsc/index.html))

**Comment by hamza1311** [comment]
https://github.com/hamza1311/reqwasm/commit/445e9a5bf555a0e37d0ff7fb71e69d34cb8f2493 should resolve the issue related with the ws unexpectedly closing. See the updated test as an example on how to use it
https://github.com/hamza1311/reqwasm/blob/445e9a5bf555a0e37d0ff7fb71e69d34cb8f2493/src/websocket/futures.rs#L300-L323 

I will resolve #11 before publishing the next release

**Comment by surt666** [comment]
Cool. Trying to make it work in my setup, but not being able to clone websocket gives all kind of lifetime issues when moving the socket between components, so have to figure out how to make that work

**Comment by hamza1311** [comment]
@surt666, was the issue resolved?

**Comment by surt666** [comment]
It seems that the close issue is resolved, but I'm somewhat unsure of how to move the socket around so it's only created in one place, now that it can't be cloned. It seems like i'm lacking some liftetime understanding, about how to move it as a reference, so if you have an example where the websocket is created outside of the function that uses it I think that would benefit a lot of people. I'm thinking taking the websocket as a parameter to the test function above

**Comment by hamza1311** [comment]
Sorry it took me a little while to respond. I recommend using channels or some other way to communicate between components inside your application instead of passing around the `WebSocket` instance. The `Clone` implementation was there to allow sending and receiving from multiple futures but that has been taken care of by `ws.split()`

**Comment by hamza1311** [comment]
Closing because this issue is resolved and 0.3.0 is now released
